src/StoreApi/Utilities/OrderAuthorizationTrait.php

```diff
@@ -0,0 +1,63 @@
+<?php
+namespace Automattic\WooCommerce\StoreApi\Utilities;
+
+use Automattic\WooCommerce\StoreApi\Utilities\OrderController;
+use Automattic\WooCommerce\StoreApi\Exceptions\RouteException;
+
+/**
+ * OrderAuthorizationTrait
+ *
+ * Shared functionality for getting order authorization.
+ */
+trait OrderAuthorizationTrait {
+	/**
+	 * Check if authorized to get the order.
+	 *
+	 * @throws RouteException If the order is not found or the order key is invalid.
+	 *
+	 * @param \WP_REST_Request $request Request object.
+	 * @return boolean|WP_Error
+	 */
+	public function is_authorized( \WP_REST_Request $request ) {
+		$order_id      = absint( $request['id'] );
+		$order_key     = wc_clean( wp_unslash( $request->get_param( 'key' ) ) );
+		$billing_email = wc_clean( wp_unslash( $request->get_param( 'billing_email' ) ) );
+
+		try {
+			// In this context, pay_for_order capability checks that the current user ID matches the customer ID stored
+			// within the order, or if the order was placed by a guest.
+			// See https://github.com/woocommerce/woocommerce/blob/abcedbefe02f9e89122771100c42ff588da3e8e0/plugins/woocommerce/includes/wc-user-functions.php#L458.
+			if ( ! current_user_can( 'pay_for_order', $order_id ) ) {
+				throw new RouteException( 'woocommerce_rest_invalid_user', __( 'This order belongs to a different customer. Please log in to the correct account.', 'woo-gutenberg-products-block' ), 403 );
+			}
+			if ( get_current_user_id() === 0 ) {
+				$this->order_controller->validate_order_key( $order_id, $order_key );
+				$this->validate_billing_email( $order_id, $billing_email );
+			}
+		} catch ( RouteException $error ) {
+			return new \WP_Error(
+				$error->getErrorCode(),
+				$error->getMessage(),
+				array( 'status' => $error->getCode() )
+			);
+		}
+
+		return true;
+	}
+
+	/**
+	 * Validate a given billing email against an existing order.
+	 *
+	 * @throws RouteException Exception if invalid data is detected.
+	 * @param integer $order_id Order ID.
+	 * @param string  $billing_email Billing email.
+	 */
+	public function validate_billing_email( $order_id, $billing_email ) {
```

Comment:
Should we improve the name of this method? `validate_billing_email` doesn't describe very well what the method does.

I suggest `check_billing_email_matches_order` or similar.

Reply:
I'm not sure that this naming makes more sense to me.  I think this falls under validation and the word "check" is a little more ambiguous.

This function was already added before this PR, but if you think we need to rename it here, we could call this `validate_billing_email_matches_order` if that's more clear?

Reply:
> we could call this `validate_billing_email_matches_order` if that's more clear?

that would be perfect!

Reply:
Updated in [a60597e](https://github.com/woocommerce/woocommerce-blocks/pull/9743/commits/a60597e0b8684a221ae4d7fe6e383901cdce0f2d).